### ISSUE TITLE
exit(1) between init() and trainingRun() (likely mem corruption)

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#94]
**Audit context:** Filed as part of a USERAPI audit of ODT (Pass 1, April 2026).
Line numbers below reference `es-ude/OnDeviceTraining@b0a737a9d1ff71d71187c6600b56ed5eff4eaab0`.

## Summary

A `snprintf` or `gmtime_r` call placed between `init()` and `trainingRun()` at the application level triggers a silent `exit(1)` at the start of ODT's training loop on macOS. No diagnostic output. Moving the string-formatting / time calls elsewhere (before `init()` or into a separate process) avoids the crash. Working hypothesis: a memory-corruption bug in an ODT static-init path whose symptom is gated on stack/dyld layout.

## Severity

| Axis | Value |
|---|---|
| Priority | Medium |
| Affected users | Host users (observed on macOS) whose application performs string formatting / time calls between `init()` and `trainingRun()` |
| Failure mode | Silent `exit(1)` (no stderr output) |
| Blast radius | Memory safety — very likely an uninitialized read or buffer overrun in ODT static-init |

## Root cause

**Not fully isolated.** Symptom-level observation: the pattern

```c
init(...);
char buf[64];
snprintf(buf, sizeof(buf), "%s", some_str);   // or gmtime_r(...)
trainingRun(...);                              // silent exit(1) starts here
```

reliably reproduces the crash on macOS during base-project development; relocating `snprintf` / `gmtime_r` to before `init()` or to an out-of-process helper avoids it. This points to a static-init-time memory issue in ODT (an uninitialized read or buffer overrun) whose visible symptom depends on how `snprintf` / `gmtime_r` perturb the stack or the dyld loader's internal state. `PRINT_ERROR` output is not available during the crash window (see the separate audit issue on `PRINT_ERROR` wiring), so no diagnostic is emitted before `exit(1)`.

## Reproducer

Minimal:

```c
#include "TrainingLoopApi.h"
#include <stdio.h>
#include <time.h>

int main(void) {
    init(/* ... */);

    char buf[64];
    time_t t = time(NULL);
    struct tm tm;
    gmtime_r(&t, &tm);
    snprintf(buf, sizeof(buf), "%04d-%02d-%02d",
             tm.tm_year + 1900, tm.tm_mon + 1, tm.tm_mday);

    trainingRun(/* ... */);   // never reached — silent exit(1)

    return 0;
}
```

Observed on macOS, default-configured host build of ODT. Move the `gmtime_r` + `snprintf` block to before `init()`: the crash disappears.

## Impact

- Any user who writes a training-run timestamp or tag between `init()` and `trainingRun()` loses the run silently — the application exits with status 1 before a single training step executes.
- On Linux or different allocators the symptom may differ; the upstream bug is likely platform-independent but the visible *failure mode* is allocator-/dyld-specific.
- The workaround ("do all string formatting before `init()` or in a helper process") has shaped the base-project's Python harness design to side-step this entirely — but that is not a fix, it is avoidance.

## Recommended fix

**Step 1:** run ODT's test suite (or a minimal reproducer) under Address Sanitizer and Undefined Behavior Sanitizer:

```bash
cmake -DCMAKE_C_FLAGS="-fsanitize=address,undefined" ...
# run any static-init-exercising test
```

Almost certainly this will surface an uninitialized read, a buffer overrun, or a use-after-free at static-init time. The fix is the one-liner that sanitizer points at.

**Step 2:** once the sanitizer pinpoints the offending line, the fix is typically a single missing initialization or a sized-buffer-overflow correction. Post-fix, re-run the reproducer above — it should complete normally.

## Verification

Reproducer above must run to completion with exit status 0. Additionally, the full ODT test suite must pass cleanly under `-fsanitize=address,undefined` with no diagnostic output from the sanitizer.

## Related

Co-located with the tracking index for this audit batch. The silent nature of the failure is compounded by the separate `PRINT_ERROR` wiring issue — fixing that first would have surfaced any attempt ODT makes to report the error before exiting.

**Comment by LeoBuron** [comment]
Confirmed fixed on `develop` via PR #131 (regression test added in `UnitTestMnistSmoke.c` reproducing the snprintf+gmtime_r-between-setup-and-trainingRun pattern; passes under both unit_test and unit_test_asan). Original symptom was nebenbei resolved by #99/#100/#101 refactors. Closing manually — develop merge does not trigger GitHub auto-close.
